ath79: small fixes for Engenius boards

mcprat opened the pull request:
### ath79: move small flash Engenius boards to tiny
for:
 - EAP350 v1
 - ECB350 v1
 - ENH202 v1

Because of the unique situation for these boards:

 - 8 MB flash
 - an extra "failsafe" image for recovery
 - TFTP does not work
 - bootloader loads image from a longer flash offset
 - 1 eraseblock each needed for OKLI kernel loader and fake rootfs
 - using mtd-concat to make use of remaining space...

factory.bin builds are already failing on master branch
so move them to the tiny subtarget

this still may not be enough for LuCI in future releases...
maybe should be handled another way in the future...
(disable LuCI?)

the user has other options such as:

flashing without LuCI and installing LuCI afterward
or
removing the "failsafe" image from the DTS
with the risk that involves (extremely difficult recovery)

---

### ath79: make Engenius fakeroot partitions read-only

for:
 - ENH202 v1
 - ENS202EXT v1

these boards were committed before it was discovered
that for all Engenius boards with a "failsafe" image
forcing the failsafe image to load next boot
can be achieved by editing the u-boot environment like so

  `fw_setenv rootfs_checksum 0`

so its not necessary to delete a partition to trigger it

now all "fake" partitions are read-only for all boards
both the fake kernel: OKLI kernel loader
and the fake rootfs: empty squashfs

---

### ath79: make all eth ports LAN for Engenius APs

for:
 - ENH202 v1
 - ENS202EXT v1
 - EnstationAC v1

on second thought...these boards are sold as
Client Bridge or Point to Point or Access Point
so there is probably no benefit to have WAN by default
for one of the ports

---

### ath79: use internal switch for ENS202EXT

Have both ports on the built-in switch of AR9341
in order to take advantage of memory and CPU savings
and switch offloading features like STP and VLAN

Also, the use of GMAC1 with internal switch for the ports
fixes the issue of the ethernet LEDs not functioning.
The LEDs are triggered by the internal switch, not a GPIO.

---

### ath79: use internal switch and small fixes for EAP300 v2

other small corrections

 - model after compatible, capitalization
 - spi frequency to 40 MHz

Have the port use GMAC1 with internal switch
which fixes the issue of the ethernet LED not functioning
The LED is triggered by the internal switch, not a GPIO.

The GPIO for the ethernet LED was added in ath79
as it was defined in the ar71xx target
but it was not functioning in ath79 for a previously unknown reason.

It is unknown why that GPIO was defined as an LED in ar71xx.

---

### ath79: catch Engenius factory.bin build error

Add logic to common-engenius.mk
so that in the event where the recipe fails
there is not an incomplete or broken factory.bin

Instead, it is removed and the build of other images
is allowed to continue.

Suggested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
ath79: make all eth ports LAN for Engenius APs

for:
 - ENH202 v1
 - ENS202EXT v1
 - EnstationAC v1
 - EWS511AP

For EWS511AP, have default behavior as static ip
to match the behavior of all other APs in ath79

These boards are sold as
Client Bridge or Point to Point or Access Point
so there is probably no benefit to have WAN by default
for one of the ports, to prevent user confusion.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

blocktrron commented on target/linux/ath79/image/tiny.mk:
```diff
@@ -1,5 +1,25 @@
+# This needs to make /tmp/_sys/sysupgrade.tgz an empty file prior to
+# sysupgrade, as otherwise it will implant the old configuration from
+# OEM firmware when writing rootfs from factory.bin
+define Build/engenius-tar-gz
```
this looks identical to the one already used for the generic subtarget? If so --> `common-engenius.mk`

mpratt14 replied:
ok will do

blocktrron commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -224,15 +224,16 @@ ath79_setup_interfaces()
 	engenius,enh202-v1)
-		ucidef_set_interface_wan "eth0"
+		ucidef_set_interface_lan "eth0"
 		ucidef_add_switch "switch0" \
 			"0@eth1" "4:lan:1"
 		;;
 	engenius,ens202ext-v1)
-		ucidef_set_interface_wan "eth1"
+		ucidef_set_interface_lan "eth1"
 		ucidef_add_switch "switch0" \
 			"0@eth0" "1:lan"
 		;;
```
At this point, you are most likely better off reconfiguring the PHY multiplexer to avoid pushing LAN traffic thru the CPU.

mpratt14 replied:
sorry this is a new idea for me...how to proceed? any examples?

mpratt14 replied:
> reconfiguring the PHY multiplexer to avoid pushing LAN traffic thru the CPU

the boards in question with `ucidef_add_switch` have no gigabit capabilities....if that is related...

I don't see any possible performance issue but I would like to know more about what you mean

blocktrron replied:
See the AR9341 datasheet on page 47 (Ethernet Subsystem). PHY0 and PHY4 can be XOR'd to the switch or the other MAC respectively. Alternatively, both PHY can be connected to the switch.

mpratt14 replied:
Yes I'm aware of this, so you mean that I should have both ethernet ports on the internal switch at GMAC1? And is there really an advantage for this (saving CPU usage as you said)?

mpratt14 replied:
If I understand correctly...

there is a benefit when both ports are used and there is traffic between the ports, otherwise traffic goes via

GMAC0 -- CPU -- GMAC1

where otherwise the internal switch handles that traffic?

blocktrron replied:
Yes, this save resources (Not having to allocate continuous memory for the second mac, less CPU interrupts for passing traffic, utilizing the switch offloading features (STP, VLAN).

mpratt14 replied:
Unless you have a suggestion for something I may be doing wrong, I will open a bug report and link here

in order to have both ports on the internal switch, it would be ports 1 and 5, but 5 is missing from swconfig on ath79
otherwise, configuration would be like this

if I understand correctly, nothing in DTS effects the number of ports you see in swconfig

```
		ucidef_add_switch "switch0" \
			"0@eth0" "1:lan" "5:lan"
```

adschm commented:
Be aware that if you move to tiny, you also have to move all mentions in base-files (if there are any)!

mpratt14 commented:
@adschm of course! I had a feeling that it was too easy just to edit the makefiles haha

updated first commit with both suggestions, rebased, edited commit messages

mpratt14 commented:
It should be good now, but I am going to flash one of them to make sure

mpratt14 commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -223,16 +221,12 @@ ath79_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0@eth0" "1:lan"
 		;;
+	engenius,enstationac-v1|\
 	engenius,ews511ap)
 		ucidef_set_interface_lan "eth0 eth1" "dhcp"
 		;;
```
a side note: is it possible that someone would have a very hard time with SSH or failsafe mode without a router since DHCP is the default protocol?

Should we consider not having DHCP as default for any board?

mpratt14 replied:
I just realized this is a bad question, but I still think it is pointless to have the "dhcp", and they should be merged with the other cases

adschm commented:
I consider merging your first patch. I'd be happy if you could improve your commit message; at the moment, it's more like a bunch of fragments than a consistent story/explanation. Making use of full sentences and corresponding capitalization might help.

mpratt14 commented:
@adschm updated commit messages (let me know if its still not good)

first 2 commits are ready, last 2 commits are not ready

mpratt14 commented:
@blocktrron one of the reasons both ports are not already on the internal switch is because I had so much trouble getting both ports working at all. However, I now have a better idea why...

I think there is a problem with the switch driver: port 5 is always missing

As I understand, ar71xx and ath79 targets use different drivers for the internal switch
(or in other words, the driver was reworked heavily and moved)

for ar71xx:
target/linux/ar71xx/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_ar7240.c

and for ath79:
target/linux/generic/files/drivers/net/phy/ar8216.c

no matter what I do in the DTS on ath79, port 5 never appears in swconfig
I have even backported the device to 19.07 branch
there is only a difference between targets, not versions

this seems to not be noticed for other devices, for example WDR3500, because GMAC0 uses what would be port 5...

ar71xx
```
root@OpenWrt:/# swconfig dev switch0 help
switch0: eth1(AR934X built-in switch), ports: 6 (cpu @ 0), vlans: 16
```

```
root@OpenWrt:/# swconfig dev switch0 port 5 show
Port 5:
        enable_mirror_rx: 0
        enable_mirror_tx: 0
        pvid: 1
        link: port:5 link:down
```

ath79
```
root@OpenWrt:/# swconfig dev switch0 help
switch0: mdio-bus.0:1f(Atheros AR8229), ports: 5 (cpu @ 0), vlans: 16
```
```
root@OpenWrt:/# swconfig dev switch0 port 5 show
Port 5:
        mib: ???
        pvid: ???
        link: ???
```

adschm commented:
I've picked the first two and will merge them later today.

mcprat committed:
ath79: use internal switch and small fixes for EAP300 v2

other small corrections

 - model after compatible, capitalization
 - spi frequency to 40 MHz

Have the port use GMAC1 with internal switch
which fixes the issue of the ethernet LED not functioning
The LED is triggered by the internal switch, not a GPIO.

The GPIO for the ethernet LED was added in ath79
as it was defined in the ar71xx target
but it was not functioning in ath79 for a previously unknown reason.

It is unknown why that GPIO was defined as an LED in ar71xx.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@blocktrron @981213 

I have added a bug report for the issue I have that is preventing me from having both ports on the switch.

I would like to have the ports on the switch on GMAC1 for both boards, as I recently discovered that it also fixes the ethernet LED activity for both ENS202EXT and EAP300 v2

https://bugs.openwrt.org/index.php?do=details&task_id=3572

adschm commented:
@mpratt14 Please rebase ...

mpratt14 commented:
@adschm rebased, also a new commit. The commit for ENS202EXT requires that the bug I mentioned gets resolved and it would break ethernet in the meantime.

981213 commented:
I remembered implementing qca,phy4-mii-enable in ar8216 back then. (Forgot whether I tested without it though.)
However it's set in SoC dtsi for all devices.
You may try this:
```
&builtin_switch {
	/delete-property/ qca,phy4-mii-enable;
};
```
to enable port 5 in the built-in switch.

mpratt14 commented:
@981213 thanks this is a big step, but not complete yet...

I can now see port 5 in swconfig, with the link status and traffic being counted, and the VLAN is properly set

however, there is no ping on port 5, I can only access on port 1

mpratt14 commented:
@981213 it is able to get the MAC for port 5 but there is no reply

mcprat committed:
ath79: catch Engenius factory.bin build error

Add logic to common-engenius.mk
so that in the event where the recipe fails
there is not an incomplete or broken factory.bin

Instead, it is removed and the build of other images
is allowed to continue.

Suggested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
Any ideas why port 5 has no traffic but has link status?

mpratt14 commented:
an odd thing I noticed with the following configuration: both ports showing link up

I would imagine that switch-phy-swap makes phy0 and phy4 switch ports and nothing else...

```
@@ -157,6 +157,7 @@

 &builtin_switch {
        /delete-property/ qca,phy4-mii-enable;
+       /delete-node/ mdio-bus;
 };

 &eth1 {
@@ -164,6 +165,7 @@

        gmac-config {
                device = <&gmac>;
+               switch-phy-swap = <1>;
                switch-only-mode = <1>;
        };
 };
```

```
Port 0:
        mib: MIB counters
RxGoodByte  : 39970 (39.0 KiB)
TxByte      : 213132 (208.1 KiB)

        pvid: 0
        link: port:0 link:up speed:1000baseT full-duplex txflow rxflow
Port 1:
        mib: MIB counters
RxGoodByte  : 108472 (105.9 KiB)
TxByte      : 13222 (12.9 KiB)

        pvid: 1
        link: port:1 link:up speed:100baseT full-duplex auto
Port 2:
        mib: No MIB data
        pvid: 0
        link: port:2 link:down
Port 3:
        mib: No MIB data
        pvid: 0
        link: port:3 link:down
Port 4:
        mib: No MIB data
        pvid: 0
        link: port:4 link:down
Port 5:
        mib: MIB counters
RxGoodByte  : 100388 (98.0 KiB)
TxByte      : 110256 (107.6 KiB)

        pvid: 1
        link: port:5 link:up speed:10baseT full-duplex auto
VLAN 1:
        vid: 1
        ports: 0t 1 5

```

mpratt14 commented:
@adschm dropping the commit for ENS202EXT until the port 5 situation is solved, so we can get this PR closed...

Maybe I should continue the conversation after branching? should I make a new bug report that is more accurate?

also added a new commit based on your suggestion from many months ago...
https://github.com/openwrt/openwrt/pull/3014#issuecomment-683864359
if you don't like the idea, you can just drop that commit

adschm commented:
What's the status for the network patches here? Are they tested and working?

Can you provide /etc/board.json and /etc/config/network for patched ens202ext-v1?
I'm not really sure whether you need to specify both eth0 and eth1 for ucidef_set_interface_lan there ...

mpratt14 commented:
@adschm I just removed the commit that was not working because of the port 5 issue, so the 3 commits here are good

because of the issue with the built-in switch, both GMAC have to be used for ENS202EXT right now

adschm commented:
Just in case you overlooked it, please paste the /etc/board.json and /etc/config/network for patched ens202ext-v1

mpratt14 commented:
@adschm I can also replicate the problem on WDR3500 with the same changes to DTS. At this point, I'm certain this is a driver issue...I have tried many things...

relevant part of board.json

```
        "switch": {
                "switch0": {
                        "enable": true,
                        "reset": true,
                        "ports": [
                                {
                                        "num": 0,
                                        "device": "eth0",
                                        "need_tag": false,
                                        "want_untag": false
                                },
                                {
                                        "num": 1,
                                        "role": "lan"
                                },
                                {
                                        "num": 5,
                                        "role": "lan"
                                }
                        ],
                        "roles": [
                                {
                                        "role": "lan",
                                        "ports": "1 5 0t",
                                        "device": "eth0.1"
                                }
                        ]
                }
        },
        "network": {
                "lan": {
                        "ifname": "eth0.1",
                        "protocol": "static"
                }
        },
        "system": {
                "label_macaddr": "88:dc:96:47:b3:36"
        }
```

network config
```
config interface 'loopback'
        option ifname 'lo'
        option proto 'static'
        option ipaddr '127.0.0.1'
        option netmask '255.0.0.0'

config globals 'globals'
        option ula_prefix 'fd14:6cf8:4802::/48'

config interface 'lan'
        option type 'bridge'
        option ifname 'eth0.1'
        option proto 'static'
        option ipaddr '192.168.1.1'
        option netmask '255.255.255.0'
        option ip6assign '60'

config switch
        option name 'switch0'
        option reset '1'
        option enable_vlan '1'

config switch_vlan
        option device 'switch0'
        option vlan '1'
        option ports '1 5 0t'
```

adschm commented:
Sure that this is from ens202ext-v1?

Because I don't see how 
```
		ucidef_set_interface_lan "eth1"
		ucidef_add_switch "switch0" \
			"0@eth0" "1:lan"
```

translates into

```
                                        "ports": "1 5 0t",
                                        "device": "eth0.1"
```

mpratt14 commented:
Remember I had removed the patch from this PR 👍🏼

mpratt14 commented:
@adschm the ens202ext patch is here in this branch, with the others, and an experiment i did

https://github.com/mpratt14/openwrt/commits/engenius-fix-switch

mpratt14 commented:
@adschm Like I said, I did the same thing with WDR3500 and it has the same problem, no traffic on port 5

https://github.com/mpratt14/openwrt/commit/1fc8bf3cde296516c5f20efadfbaab1d6c0ecc4a

adschm commented:
Can you please just test with the patches in _this_ PR and then post the requested files? My question is not at all about port 5 but about the config in 02_network which looks wrong to me.

mpratt14 commented:
@adschm to be clear, because I removed the other patch, the only change now for ens202ext is one port from WAN to LAN
(the diff makes it look like more is going on)

either way here is the output with this PRs branch, and yes both ports are tested

```
        "network": {
                "lan": {
                        "ifname": "eth1 eth0.1",
                        "protocol": "static"
                }
        },
        "switch": {
                "switch0": {
                        "enable": true,
                        "reset": true,
                        "ports": [
                                {
                                        "num": 0,
                                        "device": "eth0",
                                        "need_tag": false,
                                        "want_untag": false
                                },
                                {
                                        "num": 1,
                                        "role": "lan"
                                }
                        ],
                        "roles": [
                                {
                                        "role": "lan",
                                        "ports": "1 0t",
                                        "device": "eth0.1"
                                }
                        ]
                }
        },
```
```
config interface 'loopback'
        option ifname 'lo'
        option proto 'static'
        option ipaddr '127.0.0.1'
        option netmask '255.0.0.0'

config globals 'globals'
        option ula_prefix 'fd10:b5dd:e15f::/48'

config interface 'lan'
        option type 'bridge'
        option ifname 'eth1 eth0.1'
        option proto 'static'
        option ipaddr '192.168.1.1'
        option netmask '255.255.255.0'
        option ip6assign '60'

config switch
        option name 'switch0'
        option reset '1'
        option enable_vlan '1'

config switch_vlan
        option device 'switch0'
        option vlan '1'
        option ports '1 0t'
```

adschm commented:
Okay, I picked the two network patches. I'm not convinced yet about the specific implementation of the factory patch.
Note that I removed the unrelated SPI frequency and model property changes for the given commits. If you want to fix them, please provide separate patches.

mpratt14 commented:
@adschm ok thanks, and sorry for the confusion. Someday after more engenius products are added, I will have some patch that consolidates nodes from many DTS into a DTSI, and that would be a better time to do the SPI frequency and model edits...

the 3rd patch for the factory.bin recipe is inspired by your old comment

>  the engenius-tar-gz recipe has some room for improvement. At the moment, if the recipe fails, it will just print "Error 1 (ignored) ...", but the build won't be stopped and the already existing factory image from before the failed build step is copied into the bin directory

the fix is inspired by this

https://stackoverflow.com/questions/44337395/how-can-i-catch-an-error-flag-and-continue-the-compilation-in-makefile

if you don't like it, or you feel it's unnecessary, you can go ahead and close the PR

ath79 will migrate from swconfig to DSA at some point correct? probably that will fix the port 5 issue in the future?
I probably would open a new PR at that point... or is it going to happen soon?
